posthog/api/websocket.py

```diff
@@ -0,0 +1,45 @@
+from channels.generic.websocket import AsyncWebsocketConsumer
+from rest_framework.exceptions import Throttled
+from rest_framework.views import APIView
+from rest_framework.request import Request
+from django.contrib.auth.models import AnonymousUser
+from asgiref.sync import sync_to_async
+
+from django.test import RequestFactory
+from typing import cast
+import json
+
+
+class BaseWebsocketConsumer(AsyncWebsocketConsumer):
+    async def check_authentication(self):
+        """Check if the user is authenticated."""
+        user = self.scope.get("user", AnonymousUser())
+
+        if not user.is_authenticated:
+            await self.close(code=401)
+            return False
+        return True
+
+    async def check_throttling(self, data):
+        # Apply throttling
+        throttles = self.get_throttles(data)
```

Comment:
I am personally not a fan of referring to methods (`get_throttles`) or attributes (`scope`) not implemented in this class. I would either make these abstract methods and abstract properties to force classes to implement them, or provide a default implementation in this class.

This is safer as it forces subclasses to implement the methods/properties, so they will always be present when we try to run this. It is also simpler to read as I went all the way to the django code trying to find an implementation of `get_throttles`.